# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: widget/windows/WinUtils.cpp
# Commit: c1529fb7e0c0
# Full Hash: c1529fb7e0c07fa637e55c1991a7f1859b4f33b1
# Author: David P <daparks@mozilla.com>
# Date: 2025-07-26 21:11:15
# Description:
#   Bug 1979113: Reject AsyncEncodeAndWriteIcon promise if task dispatch fails on Windows r=daisuke,win-reviewers,gstoll
#   
#   Reject the promise at destruction unless it has already been resolved/rejected.
#   Also return an error (that is ignored).
#   
# ==============================================================================

diff -r d11795f3942b -r c1529fb7e0c0 widget/windows/WinUtils.cpp
--- a/widget/windows/WinUtils.cpp	Fri Jul 25 23:09:48 2025 +0000
+++ b/widget/windows/WinUtils.cpp	Fri Jul 25 23:26:06 2025 +0000
@@ -826,9 +826,7 @@
   nsCOMPtr<nsIRunnable> event = new AsyncEncodeAndWriteIcon(
       path, std::move(data), stride, size.width, size.height,
       aRunnable.forget(), std::move(aPromiseHolder));
-  aIOThread->Dispatch(event, NS_DISPATCH_NORMAL);
-
-  return NS_OK;
+  return aIOThread->Dispatch(event, NS_DISPATCH_NORMAL);
 }
 #endif
 
@@ -895,7 +893,11 @@
   return rv;
 }
 
-AsyncEncodeAndWriteIcon::~AsyncEncodeAndWriteIcon() {}
+AsyncEncodeAndWriteIcon::~AsyncEncodeAndWriteIcon() {
+  if (mPromiseHolder) {
+    mPromiseHolder->RejectIfExists(NS_ERROR_FAILURE, __func__);
+  }
+}
 
 AsyncDeleteAllFaviconsFromDisk::AsyncDeleteAllFaviconsFromDisk(
     bool aIgnoreRecent)
